src/transactions/TransactionUtils.cpp

```diff
@@ -1142,34 +1142,245 @@ claimableBalanceFlagIsValid(ClaimableBalanceEntry const& cb)
+
+            auto pool =
+                loadLiquidityPool(ltxInner, poolTL().asset.liquidityPoolID());
+            auto balance = poolTL().balance;
+            auto cbSponsoringAccID =
+                poolTL().ext.v() == 1 &&
+                        getLedgerEntryExtensionV1(poolShareTrustLine.current())
+                            .sponsoringID
+                    ? *getLedgerEntryExtensionV1(poolShareTrustLine.current())
+                           .sponsoringID
+                    : accountID;
+
+            // release reserves for pool share trustline that will be deleted
+            // later
+            {
+                auto trustAcc = stellar::loadAccount(ltxInner, accountID);
+                removeEntryWithPossibleSponsorship(
+                    ltxInner, ltxInner.loadHeader(),
+                    poolShareTrustLine.current(), trustAcc);
+            }
+
+            auto redeemIntoClaimableBalance =
+                [&ltxInner, &txSourceID, &txSeqNum, &opIndex, &poolTL,
+                 &accountID, &cbSponsoringAccID](Asset const& assetInPool,
+                                                 int64_t amount) -> bool {
+                // create a claimable balance for the asset in the pool
+                LedgerEntry claimableBalance;
+                claimableBalance.data.type(CLAIMABLE_BALANCE);
+
+                auto& claimableBalanceEntry =
+                    claimableBalance.data.claimableBalance();
+
+                OperationID operationID;
+                operationID.type(ENVELOPE_TYPE_POOL_REVOKE_OP_ID);
+                operationID.revokeId().sourceAccount = txSourceID;
+                operationID.revokeId().seqNum = txSeqNum;
+                operationID.revokeId().opNum = opIndex;
+                operationID.revokeId().liquidityPoolID =
+                    poolTL().asset.liquidityPoolID();
+                operationID.revokeId().asset = assetInPool;
+
+                claimableBalanceEntry.balanceID.type(
+                    CLAIMABLE_BALANCE_ID_TYPE_FROM_POOL_REVOKE);
+                claimableBalanceEntry.balanceID.fromPoolRevoke() =
+                    xdrSha256(operationID);
+
+                claimableBalanceEntry.amount = amount;
+                claimableBalanceEntry.asset = assetInPool;
+
+                ClaimPredicate u;
+                u.type(CLAIM_PREDICATE_UNCONDITIONAL);
+
+                Claimant c;
+                c.v0().destination = accountID;
+                c.v0().predicate = u;
+
+                claimableBalanceEntry.claimants = {c};
+
+                // if this asset isn't native
+                // 1. set clawback if it's set on the trustline
+                // 2. decrement liquidityPoolUseCount on the asset trustline
+                if (assetInPool.type() != ASSET_TYPE_NATIVE)
+                {
+                    // trustline must exist
+                    auto assetTrustLine =
+                        ltxInner.load(trustlineKey(accountID, assetInPool));
```

Comment:
Nitpick: can we use `loadTrustLine` here?

Reply:
Not without making some other changes. We need to access the extension to modify `liquidityPoolUseCount`.

Related to this though, there's a bug here. If `accountID` is the issuer of `assetInPool`, then this will throw. This is possible because the other asset (not the one being revoked) can be issued by `accountID`. This is easy to fix, but the interesting question to ask here is should we bother creating the claimable balance in this case? I don't think we should since the claimable balance will just be "burned" when it's claimed. Either way, I'll update the CAP to clarify this detail once we decide what to do.

Reply:
I updated the implementation to not create the claimable balance if the claimant is the issuer. I'll also update the CAP and post the link to the PR here.

Reply:
https://github.com/stellar/stellar-protocol/pull/1042